Migrate to declarative validation: ReplicationController spec.replicas and spec.minReadySeconds fields

jpbetz opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This migrates the ReplicationController `spec.replicas` and `spec.minReadySeconds` fields to declarative validation:
  - Add `+k8s:minimum` declarative validation for both fields.
  - Mark the fields as optional using `+k8s:optional` and set a declarative default with `+default` so that declarative validation understands the requiredness  of the fields.  Note that since both fields are optional but also have a default, they are "required" from a server side validation perspective (validation happens after defaulting).
- Add versioned validation testing in `declarative_validation_test.go`. Declarative validation is performed on the versioned types, and the test infrastructure automatically compares validation error results of handwritten validation against versioned declarative validation.

#### Does this PR introduce a user-facing change?

```release-note
The minimum value validation of ReplicationController's `replicas` and `minReadySeconds` fields have been migrated to declarative validation. The requiredness of both fields is also declaratively validated.
If the `DeclarativeValidation` feature gate is enabled, mismatches with existing validation are reported via metrics.
If the `DeclarativeValidationTakeover` feature gate is enabled, declarative validation is the primary source of errors for migrated fields.
```

thockin committed:
Declaratively validate minimum value of RC.Spec.Replicas

thockin committed:
Add declarative default for RC.Spec.Replicas

Remove manual default:

```
$ git diff
diff --git a/pkg/apis/core/v1/defaults.go b/pkg/apis/core/v1/defaults.go
index e66de8bb432..1dd28dd35fb 100644
--- a/pkg/apis/core/v1/defaults.go
+++ b/pkg/apis/core/v1/defaults.go
@@ -60,10 +60,6 @@ func SetDefaults_ReplicationController(obj *v1.ReplicationController) {
                        obj.Labels = labels
                }
        }
-       if obj.Spec.Replicas == nil {
-               obj.Spec.Replicas = new(int32)
-               *obj.Spec.Replicas = 1
-       }
 }
 func SetDefaults_Volume(obj *v1.Volume) {
        if ptr.AllPtrFieldsNil(&obj.VolumeSource) {
```

The test fails:

```
$ go test ./pkg/apis/core/v1 | grep -v gate | grep -v SetEmulationVersion
--- FAIL: TestSetDefaultReplicationControllerReplicas (0.00s)
    defaults_test.go:1608: expected: 1 replicas, got: 0
FAIL
FAIL	k8s.io/kubernetes/pkg/apis/core/v1	0.269s
FAIL
```

Declare the default, update codegen and re-run the test:

```
$ git diff
diff --git a/staging/src/k8s.io/api/core/v1/types.go b/staging/src/k8s.io/api/core/v1/types.go
index 406ab56a002..7e5136fe9f6 100644
--- a/staging/src/k8s.io/api/core/v1/types.go
+++ b/staging/src/k8s.io/api/core/v1/types.go
@@ -5101,6 +5101,7 @@ type ReplicationControllerSpec struct {
        // Defaults to 1.
        // More info: https://kubernetes.io/docs/concepts/workloads/controllers/replicationcontroller#what-is-a-replicationcontroller
        // +optional
+       // +default=1
        // +k8s:minimum=0
        Replicas *int32 `json:"replicas,omitempty" protobuf:"varint,1,opt,name=replicas"`

$ ./hack/update-codegen.sh default
+++ [1219 08:58:43] Generating defaulter code for 102 targets

$ git diff
diff --git a/pkg/apis/core/v1/zz_generated.defaults.go b/pkg/apis/core/v1/zz_generated.defaults.go
index 3b6eb4f0a93..567c49053aa 100644
--- a/pkg/apis/core/v1/zz_generated.defaults.go
+++ b/pkg/apis/core/v1/zz_generated.defaults.go
@@ -878,6 +878,10 @@ func SetObjectDefaults_PodTemplateList(in *corev1.PodTemplateList) {

 func SetObjectDefaults_ReplicationController(in *corev1.ReplicationController) {
        SetDefaults_ReplicationController(in)
+       if in.Spec.Replicas == nil {
+               var ptrVar1 int32 = 1
+               in.Spec.Replicas = &ptrVar1
+       }
        if in.Spec.Template != nil {
                SetDefaults_PodSpec(&in.Spec.Template.Spec)
                for i := range in.Spec.Template.Spec.Volumes {

$ go test ./pkg/apis/core/v1 | grep -v gate | grep -v SetEmulationVersion
ok  	k8s.io/kubernetes/pkg/apis/core/v1	(cached)
```

thockin committed:
Declaratively validate RC.Spec.Replicas optionality

The existing test run both declarative and manual validation and it
still passes.

thockin committed:
Add declarative default for RC.Spec.MinReadySeconds

thockin committed:
Declaratively validate minimum value of RC.Spec.MinReadySeconds

# Conflicts:
#	staging/src/k8s.io/api/core/v1/types.go

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/openapi.go:
```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2024 The Kubernetes Authors.
```
This file was broken up in dev branch - do you want to do that here or later?

jpbetz replied:
Resolving this in https://github.com/kubernetes/kubernetes/pull/130730

jpbetz commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -16659,8 +16666,31 @@ func TestValidateReplicationControllerUpdate(t *testing.T) {
 		tc.update.ObjectMeta.ResourceVersion = "1"
-		if errs := ValidateReplicationControllerUpdate(&tc.update, &tc.old, PodValidationOptions{}); len(errs) != 0 {
-			t.Errorf("expected success: %v", errs)
+		for _, gateVal := range []bool{false, true} {
```
It was just added recently after @yongruilin pointed out it was missing: https://github.com/kubernetes/kubernetes/pull/130725/commits/697decb9048e6ff7d1fd83daaac36587746a7ca4#diff-883b42cb2642e4c4b94cdb94f18c1c62c3992d62b3dfa5309f9ec569e059dfd3

yongruilin replied:
This is also reverted in dev branch https://github.com/jpbetz/kubernetes/commit/1086bc8e49c41a55d721fd759f7889e7727801bc
I think we need to merge it together and squash the commits?

jpbetz replied:
Done

jpbetz commented on pkg/apis/core/validation/validation.go:
```diff
@@ -6297,10 +6297,12 @@ func ValidateReplicationControllerSpec(spec, oldSpec *core.ReplicationController
-		allErrs = append(allErrs, field.Required(fldPath.Child("replicas"), ""))
-	} else {
-		allErrs = append(allErrs, ValidateNonnegativeField(int64(*spec.Replicas), fldPath.Child("replicas")).MarkCoveredByDeclarative()...)
+	if !utilfeature.DefaultFeatureGate.Enabled(features.DeclarativeValidation) {
```
Good catch, this snuck past me when I was forklifting the commits onto master.  I'll drop this feature gate check.

yongruilin replied:
Yea, there are 2 commits in PR should be picked https://github.com/jpbetz/kubernetes/pull/121 , the "revert" one and the "strategy_test" one

jpbetz replied:
Ok, both commits are fully picked up now.

aaron-prindle commented:
/triage accepted

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/runtime/testing/validation.go:
```diff
@@ -0,0 +1,119 @@
+//	      errs = append(errs, versionValidationErrors...) // generated declarative validation
+//		  // Validate that the errors are what was expected for this test case.
+//		})
+func RunValidationForEachVersion(t *testing.T, scheme *runtime.Scheme, options sets.Set[string], unversioned runtime.Object, fn VersionValidationRunner) {
```
Thanks. I've moved these function into https://github.com/kubernetes/kubernetes/pull/130724 and combined them with the commit with the testing that uses them.

jpbetz commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -16970,6 +16985,107 @@ func TestValidateReplicationController(t *testing.T) {
+	}
+}
+
+func verifyVersionedValidationEquivalence(t *testing.T, obj, old k8sruntime.Object) {
```
I've moved this into a testing commit in https://github.com/kubernetes/kubernetes/pull/130724.

deads2k commented on pkg/apis/core/types.go:
```diff
@@ -4332,7 +4332,7 @@ type PodTemplateList struct {
 type ReplicationControllerSpec struct {
 	// Replicas is the number of desired replicas.
-	Replicas int32
+	Replicas *int32
```
Any idea how many internal objects we're going to have to change?  Not an object, not asking for a count if we don't have one, but asking to brace myself for conversion reviews.

jpbetz replied:
I talked a bot into writing a diff script for me.  It's found:

19 int/string cases.  10 intOrString cases.

```
 ReplicationControllerSpec.Replicas: int32 -> *int32
 ServiceAccountTokenProjection.ExpirationSeconds: int64 -> *int64
 CertificateSigningRequestSpec.SignerName: string -> *string
 ServiceReference.Port: int32 -> *int32
 ServiceReference.Port: int32 -> *int32
 DeploymentSpec.Replicas: int32 -> *int32
 ReplicaSetSpec.Replicas: int32 -> *int32
 RollingUpdateDaemonSet.MaxSurge: intstr.IntOrString -> *intstr.IntOrString
 RollingUpdateDaemonSet.MaxUnavailable: intstr.IntOrString -> *intstr.IntOrString
 RollingUpdateDeployment.MaxSurge: intstr.IntOrString -> *intstr.IntOrString
 RollingUpdateDeployment.MaxUnavailable: intstr.IntOrString -> *intstr.IntOrString
 RollingUpdateStatefulSetStrategy.Partition: int32 -> *int32
 StatefulSetSpec.Replicas: int32 -> *int32
 StatefulSetStatus.ObservedGeneration: *int64 -> int64
 DeploymentSpec.Replicas: int32 -> *int32
 RollingUpdateDeployment.MaxSurge: intstr.IntOrString -> *intstr.IntOrString
 RollingUpdateDeployment.MaxUnavailable: intstr.IntOrString -> *intstr.IntOrString
 RollingUpdateStatefulSetStrategy.Partition: int32 -> *int32
 StatefulSetSpec.Replicas: int32 -> *int32
 DeploymentSpec.Replicas: int32 -> *int32
 ReplicaSetSpec.Replicas: int32 -> *int32
 RollingUpdateDaemonSet.MaxSurge: intstr.IntOrString -> *intstr.IntOrString
 RollingUpdateDaemonSet.MaxUnavailable: intstr.IntOrString -> *intstr.IntOrString
 RollingUpdateDeployment.MaxSurge: intstr.IntOrString -> *intstr.IntOrString
 RollingUpdateDeployment.MaxUnavailable: intstr.IntOrString -> *intstr.IntOrString
 RollingUpdateStatefulSetStrategy.Partition: int32 -> *int32
 StatefulSetSpec.Replicas: int32 -> *int32
 StatefulSetStatus.ObservedGeneration: *int64 -> int64
 TokenRequestSpec.ExpirationSeconds: int64 -> *int64
 ```
 
 Not sure if the intOrString matter or not.

jpbetz replied:
A bunch are literally the replicas field in different locations.

thockin replied:
The replicas pattern looks like it was cargo-culted.  Sorry, that was probably my fault.

deads2k commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -60,10 +60,7 @@ func SetDefaults_ReplicationController(obj *v1.ReplicationController) {
-		obj.Spec.Replicas = new(int32)
-		*obj.Spec.Replicas = 1
-	}
+	// obj.Spec.Replicas is defaulted declaratively
```
apologies for the dense question, but could I get a link?  I don't see it in https://github.com/kubernetes/kubernetes/blob/85488b5f10bc6128dab69e4bdd94cec2c005c7f5/pkg/apis/core/v1/zz_generated.defaults.go#L879

jpbetz replied:
From the [below comment](https://github.com/kubernetes/kubernetes/pull/130725/files#r1991378264), looks like you found it?

deads2k commented on pkg/apis/core/v1/zz_generated.defaults.go:
```diff
@@ -878,6 +878,10 @@ func SetObjectDefaults_PodTemplateList(in *corev1.PodTemplateList) {
 
 func SetObjectDefaults_ReplicationController(in *corev1.ReplicationController) {
 	SetDefaults_ReplicationController(in)
+	if in.Spec.Replicas == nil {
```
well, at least I wasn't crazy when I was checking

deads2k commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5107,12 +5107,17 @@ type ReplicationControllerSpec struct {
 	// without any of its container crashing, for it to be considered available.
 	// Defaults to 0 (pod will be considered available as soon as it is ready)
 	// +optional
+	// +default=0
```
Now this is interesting.  @thockin in a theory of declaring a marker and that marker must be zero value, what does a default of the zero value mean?  For built-ins, this won't even impact serialization. Why isn't `+k8s:optional` more appropriate

jpbetz replied:
> Now this is interesting. @thockin in a theory of declaring a marker and that marker must be zero value, what does a default of the zero value mean? 

It just documents the behavior, I think?  I kinda like the explicitness of it-- without having to grok the field's type and the zero-valued behavior, this tells a user what to expect when they do not provide a value.

> For built-ins, this won't even impact serialization. Why isn't `+k8s:optional` more appropriate

@thockin, based on the [below comment](https://github.com/kubernetes/kubernetes/pull/130725#discussion_r1992473053), should we have `k8s:optional` here given that we have `optional`?

deads2k commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5107,12 +5107,17 @@ type ReplicationControllerSpec struct {
 	// Defaults to 1.
 	// More info: https://kubernetes.io/docs/concepts/workloads/controllers/replicationcontroller#what-is-a-replicationcontroller
 	// +optional
+	// +k8s:optional
+	// +default=1
```
I definitely prefer this construct, with an explicit `+k8s:optional`.  Will we choose this as our recommended (possibly linted?) phrasing?

yongruilin replied:
Sure, we can follow up to extend the lint rules to cover if we settle with this as the best practice.

thockin replied:
I don't want to mislead - my goal is that 100% of fields which say `+optional` also say `+k8s:optional`, so that the obvious followup is to delete one in favor of the other.

If we find places where that is not true, I think we will have screwed up.  In the interim we keep both as part of our promise of being easy to back out if things go sideways.

deads2k reviewed (commented):
I'm not finished, but I have to take a pause for a few hours and want to give as much chance for answers as possible.

jpbetz committed:
Add declarative test cases for RC.Spec.Replicas

jpbetz committed:
Add declarative test cases for RC.Spec.MinReadySeconds

jpbetz committed:
Mark RC.Spec.MinReadySeconds as k8s:optional

+k8s:optional should be used everywhere +optional is.
This does not change the vailidation of the field, but it
is a good practice, and code generator recognizes the fields
tags and outputs this comment above the field validation
as a result: "optional value-type fields with zero-value
defaults are purely documentation".

deads2k commented on pkg/apis/core/v1/zz_generated.validations.go:
```diff
@@ -0,0 +1,112 @@
+	errs = append(errs,
+		func(fldPath *field.Path, obj, oldObj *int32) (errs field.ErrorList) {
+			errs = append(errs, validate.Minimum(ctx, op, fldPath, obj, oldObj, 0)...)
+			// optional value-type fields with zero-value defaults are purely documentation
```
nit: to match the required stanza doc, can this be at the top of the block?

jpbetz replied:
WDYT @thockin ?  I can't remember if the comment output can be adapted for stuff like this easily.

thockin replied:
This is an implementation wart.  Currently it tracks Comments and Functions as separate things, and when we emit the code we emit calls to validators before comments.  We could do comments first, but that isn't obviously correct either.

I thought about, but didn't do yet, making comments a special category of Function which doesn't emit an actual function call, but would still be sorted with other function calls.  It's a little ugly because the same type is used in a few places where just the comments would not make sense.

Emitting comments first is easy and PROBABLY makes sense in most cases., I can send that as a followup?

thockin replied:
Last commit in https://github.com/kubernetes/kubernetes/pull/130795 builds on this PR

deads2k commented on pkg/registry/core/replicationcontroller/declarative_validation_test.go:
```diff
@@ -0,0 +1,247 @@
+		APIGroup:   "",
+		APIVersion: "v1",
+	})
+	testCases := map[string]struct {
```
missing nil replicas?

jpbetz replied:
Since tests that use data literals to define objects have defaulting applied, are we imaging that (a) the tests fail because the field is unset or (b) that we run the defaulter and show that validation passes after defaulting?

jpbetz replied:
SG, let me add the nil cases.

thockin replied:
Joe, I added this case in the commit I sent you after you broke testcases out, too :)

jpbetz replied:
Incorporated into the commits of this PR and pushed.  Thanks!

deads2k reviewed (changes requested):
minor changes requested.  I remain concerned about how declarative validateUpdate appears to enforce validation on existing field values as I noted in the previous PR.  It appears to raise the stakes on review of every field we add declarative validation to since it's all novel.

jpbetz commented on pkg/apis/core/fuzzer/fuzzer.go:
```diff
@@ -119,6 +119,12 @@ var Funcs = func(codecs runtimeserializer.CodecFactory) []interface{} {
 			c.FillNoCustom(j) // fuzz self without calling this function again
 			//j.TemplateRef = nil // this is required for round trip
+
+			// match defaulting
+			if j.Replicas == nil {
+				replicas := int32(0)
+				j.Replicas = &replicas
+			}
```
We have defaulting registered into the scheme.  I haven't though about how to incorporate it into the fuzzer, but that sounds like a nice win.

deads2k replied:
If this is difficult to do, I can agree to this as non-blocking for 1.33, but I think we should get it early in 1.34 so that future testing is easier for us.

jpbetz replied:
For tracking: https://github.com/kubernetes/kubernetes/issues/130791

jpbetz commented on pkg/apis/core/v1/zz_generated.validations.go:
```diff
@@ -0,0 +1,112 @@
+			return
+		}(fldPath.Child("spec"), &obj.Spec, safe.Field(oldObj, func(oldObj *corev1.ReplicationController) *corev1.ReplicationControllerSpec { return &oldObj.Spec }))...)
+
+	// field corev1.ReplicationController.Status has no validation
```
This is a good catch.

We haven't put a ton of thought into subresources in 1.33.  There are a lot of considerations when you look at `*/scale`, the inconsistent metadata access of `*/status`, the complex masking of `pod/ephemeralContainers` and `pod/resize`.

I don't think we can fully address it in 1.33.  But we shouldn't migrate any subresource validation until we've got it sorted out fully.

deads2k replied:
> I don't think we can fully address it in 1.33. But we shouldn't migrate any subresource validation until we've got it sorted out fully.

Ok.  Please be careful not to forget it.  With that agreement I consider my comment non-blocking.

jpbetz replied:
SG. I'll add a guard in the code and we'll get a KEP update out at the end of 1.33 here to track all the gaps we need to address going into 1.34.

Really appreciate the attention to detail on the review.

thockin replied:
Yeah, capturing slack discussion:  It's not clear that validating across subresources is always right or always wrong.  We need to go looking for cases to see what we are currently doing.  It would be nice to bound it to a single subresource, but I'll be that somewhere we validate a staus field only if a spec field is set.

jpbetz commented:
PSA: https://github.com/kubernetes/kubernetes/pull/130732 contains the changes in the 1st commit (non-ptr -> ptr change) of this PR.  Once it merges I'll rebase this to drop the 1st commit.

jpbetz commented:
Rebased on master.

jpbetz commented:
@deads2k @thockin All feedback I'm aware of has been applied.  Please let me know if there is anything else I can address in this PR.

jpbetz commented:
/retest

deads2k commented:
>This looks as tight as it is going to get :)
>LGTM, but let David do the honors

:)

/lgtm
/approve

liggitt commented:
I think this is breaking e2e jobs which enable alpha features and enable declarative validation - like https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128152/pull-kubernetes-e2e-kind-alpha-beta-features/1900570126353174528

jpbetz commented:
> I think this is breaking e2e jobs which enable alpha features and enable declarative validation - like [prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128152/pull-kubernetes-e2e-kind-alpha-beta-features/1900570126353174528](https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/128152/pull-kubernetes-e2e-kind-alpha-beta-features/1900570126353174528)

Yes, see https://github.com/kubernetes/kubernetes/pull/130819

tengqm commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5107,12 +5107,18 @@ type ReplicationControllerSpec struct {
 	// Defaults to 1.
```
With the default value expressed declarative, we can safely remove this comment?

tengqm commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -5107,12 +5107,18 @@ type ReplicationControllerSpec struct {
 
 	// Minimum number of seconds for which a newly created pod should be ready
 	// without any of its container crashing, for it to be considered available.
 	// Defaults to 0 (pod will be considered available as soon as it is ready)
```
With the default value expressed declarative, we can safely remove this comment?

thockin replied:
I don't think the docs generation inserts the declared default into the docs, does it?  It should.
